src/python/zquantum/core/wip/circuit/gates/_gate.py

```diff
@@ -0,0 +1,388 @@
+
+        return f"{type(self).__name__}({', '.join(args_reprs)})"
+
+    def __repr__(self):
+        return str(self)
+
+
+class CustomGate(Gate):
+    """Gate class with custom matrix."""
+
+    def __init__(
+        self, matrix: sympy.Matrix, qubits: Tuple[int, ...], name: Optional[str] = None
+    ):
+        """Initialize a gate
+
+        Args:
+            matrix (sympy.Matrix): See class definition
+            qubits (tuple(int)): See class definition
+        """
+        if not self.is_valid_operator(matrix, qubits):
+            raise ValueError(
+                f"Passed matrix has shape {matrix.shape} and cannot act on "
+                f"{len(qubits)} qubits."
+            )
+
+        super().__init__(qubits)
+        copied_matrix = copy.deepcopy(matrix)
+
+        self.name = name if name is not None else str(matrix)
+
+        for index, element in enumerate(copied_matrix):
+            copied_matrix[index] = element.evalf()
+
+        self._matrix = copied_matrix
+
+    @property
+    def matrix(self) -> sympy.Matrix:
+        """Matrix representation of this gate."""
+        return self._matrix
+
+    def evaluate(self, symbols_map: Dict[str, Any]) -> "Gate":
+        """Return a copy of self with symbolic parameters substituted according to provided map.
+
+        Args:
+            symbols_map (Dict): A map of the symbols/gate parameters to new values
+
+        Returns:
+            A copy of self with the same matrix, but every symbolic param in the map is
+            substituted with corresponding value.
+        """
+        if not self.is_parameterized:
+            warnings.warn(
+                """Gate is not parameterized. evaluate will return a copy of the current gate."""
+            )
+            return copy.deepcopy(self)
+
+        if any(symbol not in self.symbolic_params for symbol in symbols_map):
+            warnings.warn(
+                f"""
+                Trying to evaluate gate with symbols not existing in the gate:
+                Symbols in circuit: {self.symbolic_params}
+                Symbols in the map: {symbols_map.keys()}
+                """,
+                Warning,
+            )
+
+        evaluated_matrix = copy.deepcopy(self.matrix)
+
+        for index, element in enumerate(evaluated_matrix):
+            new_element = element.subs(symbols_map).evalf()
+            if isinstance(sympy.re(new_element), sympy.Number) and isinstance(
+                sympy.im(new_element), sympy.Number
+            ):
+                new_element = complex(
+                    round(sympy.re(new_element), 16), round(sympy.im(new_element), 16)
+                )
+            evaluated_matrix[index] = new_element
+
+        evaluated_gate = CustomGate(evaluated_matrix, self.qubits)
+        return evaluated_gate
+
+    def __repr__(self) -> str:
+        return f"zquantum.core.circuit.gate.CustomGate(matrix={self.matrix}, qubits={self.qubits})"
+
+
+class SpecializedGate(Gate):
```

Comment:
The `Gate` base class provides the default implementation of several methods that are common to `Specialized` and `CustomGate` classes, e.g. dagger. After we rework custom gates this might no longer be the case though.

Reply:
Ok, we'll see :)